### PR TITLE
feat(funding-rate-store) Permission `getFundingRate` method, make funding rate structs private

### LINKED ISSUE #2040: Funding Rate Store: Use Registry to permission funding rate request/access
**Why**

Project Evergreen.

**Difficulty Score [1-10]**

- 4: ~2 hours

### PR DESCRIPTION
**Pull request by nicholaspai**
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

**Motivation**

#2040

**Summary**

`fundingRateRecords` and `fundingRateDisputes` mappings are now private so that funding rate data is only available via the `getFundingRateForContract()` method permissioned call.

Proposers still are expected to be interest in their a perpetual contract's base reward rate (per second) and its last update time, which they can use to project their rewards. So 1 view method is added:
- `getRewardRateForContract(perpetualAddress, proposedRate)` --> returns the projected reward rate for `perpetual` using the time elapsed since the last proposal and the magnitude of change between the `proposedRate` and the current rate.

**Issue(s)**

Fixes #2040

### CODE REVIEW

File: packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol [hunk 1]
```diff
@@ -179,6 +191,41 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
...
+     * of change between the proposed and current funding rate.
+     * @param perpetual perpetual contract whose base reward rate the caller is querying.
+     * @return FixedPoint.Unsigned representing the base reward rate for the given contract. 0.01 would represent a
+     * reward rate of 1% per second.
+     */
+    function getRewardRateForContract(address perpetual)
+        external
+        view
+        nonReentrantView()
+        returns (FixedPoint.Unsigned memory)
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol:223]: Just curious: what do you think the advantages/disadvantages would be for combining these two methods into one that returns the projected reward? (not suggesting that's what we should do -- just want to hear your thoughts)

**Review by mrice32 — COMMENTED** [review]
Looks awesome! One minor comment.

**Comment by nicholaspai** [comment]
@UMAprotocol/eng Do we also need a similar method for the `cumulativeFeeMultiplier`? Currently this has not been needed with the EMP because regular fees are 0, but wouldn't sponsors need to know their up to date fee-adjusted collateral?

File: packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol (hunk 1 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol:223]: Interesting. That could be a good idea if we anticipate that `getRewardRate` and `getProposalTime` would only be used to calculate potential rewards. Do you think there'd be other reasons to call them?

Comment (mrice32) [reply packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol:223]: Yeah, not sure. I can't think of any other use case.

**Comment by mrice32** [comment]
> @UMAprotocol/eng Do we also need a similar method for the `cumulativeFeeMultiplier`? Currently this has not been needed with the EMP because regular fees are 0, but wouldn't sponsors need to know their up to date fee-adjusted collateral?

Isn't that what `getCollateral()` does?

**Comment by nicholaspai** [comment]
> > @UMAprotocol/eng Do we also need a similar method for the `cumulativeFeeMultiplier`? Currently this has not been needed with the EMP because regular fees are 0, but wouldn't sponsors need to know their up to date fee-adjusted collateral?
> 
> Isn't that what `getCollateral()` does?

Not quite though `getCollateral` uses the current `cumulativeFeeMultiplier` but it doesn't actually update the `cumulativeFeeMultiplier`. This would return a slightly higher value if the last feeMultiplier was updated >= 1 second ago, and in the mean-time regular fees are still accruing but just haven't been reflected on the feeMultiplier.

File: packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol (hunk 1 above)
Comment (nicholaspai) [reply packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol:223]: For the sake of reducing interface size, I'll combine these for now into one method that returns projected rewards.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Comment by mrice32** [comment]
> > > @UMAprotocol/eng Do we also need a similar method for the `cumulativeFeeMultiplier`? Currently this has not been needed with the EMP because regular fees are 0, but wouldn't sponsors need to know their up to date fee-adjusted collateral?
> > 
> > 
> > Isn't that what `getCollateral()` does?
> 
> Not quite though `getCollateral` uses the current `cumulativeFeeMultiplier` but it doesn't actually update the `cumulativeFeeMultiplier`. This would return a slightly higher value if the last feeMultiplier was updated >= 1 second ago, and in the mean-time regular fees are still accruing but just haven't been reflected on the feeMultiplier.

I see. Yeah, good point! That could be nice, although, it seems annoying to compute.

**Comment by nicholaspai** [comment]
> > > > @UMAprotocol/eng Do we also need a similar method for the `cumulativeFeeMultiplier`? Currently this has not been needed with the EMP because regular fees are 0, but wouldn't sponsors need to know their up to date fee-adjusted collateral?
> > > 
> > > 
> > > Isn't that what `getCollateral()` does?
> > 
> > 
> > Not quite though `getCollateral` uses the current `cumulativeFeeMultiplier` but it doesn't actually update the `cumulativeFeeMultiplier`. This would return a slightly higher value if the last feeMultiplier was updated >= 1 second ago, and in the mean-time regular fees are still accruing but just haven't been reflected on the feeMultiplier.
> 
> I see. Yeah, good point! That could be nice, although, it seems annoying to compute.

Yeah it's not a clean refactor but without the method, sponsor positions will read collateral incorrectly via `getCollateral()` once regular fees are > 0. For now I'll leave it but add in a comment to check this out later

**Pull request merged**